### PR TITLE
US-2558: Added rules for the password creation.

### PR DESCRIPTION
**Pull request by DeepikaDK001**

### CODE REVIEW

File: src/NHSD.GPIT.BuyingCatalogue.Services/Users/UsersService.cs [hunk 1]
```diff
@@ -163,5 +171,21 @@ public async Task<bool> IsAccountManagerLimit(int organisationId, int userId = 0
 
             return users >= accountManagementSettings.MaximumNumberOfAccountManagers;
         }
+
+        public async Task<bool> IsPasswordPresentInPastNPasswords(AspNetUser user, string email, string newPassword, int numOfPreviousPasswords)
```
Comment (ghost) [inline src/NHSD.GPIT.BuyingCatalogue.Services/Users/UsersService.cs]: `numOfPreviousPasswords` could be turned into a configuration value stored in the appsettings.json file and injected in the constructor, since it should be a single value throughout the app.

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Identity/Models/ResetPasswordViewModel.cs [hunk 2]
```diff
@@ -24,6 +24,11 @@ public static class ErrorMessages
         {
             public const string PasswordRequired = "Enter a password";
             public const string PasswordMismatch = "Passwords do not match";
+            public const string PasswordPriviouslyUsed = "Password was used previously.Enter a different password";
+        }
+        public static class ConfigPasswordResetValues
+        {
+            public const int numOfPreviousPasswords = 5;
```
Comment (ghost) [inline src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Identity/Models/ResetPasswordViewModel.cs]: Can be mapped as a configuration value in appSettings.json as it isn't really a view model concern. There are some existing examples in code of mapping appSettings to objects - see `ServiceCollectionExtensions` in the WebApp.

File: src/NHSD.GPIT.BuyingCatalogue.Services/Users/UsersService.cs [hunk 3]
```diff
@@ -163,5 +171,21 @@ public async Task<bool> IsAccountManagerLimit(int organisationId, int userId = 0
...
+        {
+            var passwordHistQuery = (from hist in dbContext.AspNetUsers.TemporalAll()
```
Comment (ghost) [inline src/NHSD.GPIT.BuyingCatalogue.Services/Users/UsersService.cs]: This could be converted to the LINQ method style syntax, such as `.Where`, `.OrderByDescending`, `.Select` and `.Take`

**Review by bjss-colin — CHANGES_REQUESTED** [review]

File: tests/NHSD.GPIT.BuyingCatalogue.Services.UnitTests/Users/UsersServiceTests.cs [hunk 4]
```diff
@@ -24,7 +24,7 @@ public static class UsersServiceTests
         [Fact]
         public static void Constructor_NullRepository_ThrowsException()
         {
-            Assert.Throws<ArgumentNullException>(() => _ = new UsersService(null, new AccountManagementSettings()));
+            Assert.Throws<ArgumentNullException>(() => _ = new UsersService(null, new AccountManagementSettings(), null, null, null));
```
Comment (ghost) [inline tests/NHSD.GPIT.BuyingCatalogue.Services.UnitTests/Users/UsersServiceTests.cs]: Have a look at the `VerifyGuardClause` tests that exist in the unit test suite. It would be good to switch to that format to ensure all constructor parameters are tested

File: tests/NHSD.GPIT.BuyingCatalogue.Framework.UnitTests/Identity/PasswordValidatorTests.cs [hunk 5]
```diff
@@ -8,10 +8,7 @@ namespace NHSD.GPIT.BuyingCatalogue.Framework.UnitTests.Identity
     public static class PasswordValidatorTests
     {
         [Theory]
-        [InlineData("Pass123123")]
```
Comment (ghost) [inline tests/NHSD.GPIT.BuyingCatalogue.Framework.UnitTests/Identity/PasswordValidatorTests.cs:11]: Let's add these as a test scenario to test the negative path where the specified password does not meet the password policy requirements

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Identity/Controllers/AccountController.cs [hunk 6]
```diff
@@ -145,9 +149,15 @@ public async Task<IActionResult> ResetPassword(ResetPasswordViewModel viewModel)
         {
             if (!ModelState.IsValid)
                 return View(viewModel);
+            var user = await userManager.FindByEmailAsync(viewModel.Email);
+            var usedPassword = services.IsPasswordPresentInPastNPasswords(user, viewModel.Email, viewModel.Password);
+            if (usedPassword)
```
Comment (ghost) [inline src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Identity/Controllers/AccountController.cs:155]: This code path needs to be unit tested

File: src/NHSD.GPIT.BuyingCatalogue.Services/Users/UsersService.cs [hunk 7]
```diff
@@ -163,5 +174,22 @@ public async Task<bool> IsAccountManagerLimit(int organisationId, int userId = 0
 
             return users >= accountManagementSettings.MaximumNumberOfAccountManagers;
         }
+
+        public bool IsPasswordPresentInPastNPasswords(AspNetUser user, string email, string newPassword)
```
Comment (ghost) [inline src/NHSD.GPIT.BuyingCatalogue.Services/Users/UsersService.cs]: This method needs to be unit tested to cover the happy and unhappy path

Comment (ghost) [inline src/NHSD.GPIT.BuyingCatalogue.Services/Users/UsersService.cs]: Might be worth renaming this method now that the number of previous passwords is a config value. `IsPasswordValid` or something along those lines.

File: src/NHSD.GPIT.BuyingCatalogue.Services/Users/UsersService.cs [hunk 8]
```diff
@@ -163,5 +174,22 @@ public async Task<bool> IsAccountManagerLimit(int organisationId, int userId = 0
...
+        {
+            var passwordHistQuery1 = dbContext.AspNetUsers.TemporalAll()
+                .Where(x => x.Email == email)
+                .OrderByDescending(x => x.LastUpdated)
+                .Select(x => x.PasswordHash)
```
Comment (ghost) [inline src/NHSD.GPIT.BuyingCatalogue.Services/Users/UsersService.cs]: Having thought of this more, we will need a Distinct on the `PasswordHash`.

The Temporal tables will be updated any time a change is made to a user, including the first and/or last name. If somebody updates the name of an account 5 times it'll retrieve their **current** password five times. This will then allow the user to use their **previous** password.

As a result of this, what we should do is retrieve the last 5 _unique_ passwords and compare against those.

**Review by bjss-colin — CHANGES_REQUESTED** [review]

File: src/NHSD.GPIT.BuyingCatalogue.Services/Users/UsersService.cs [hunk 9]
```diff
@@ -163,5 +174,23 @@ public async Task<bool> IsAccountManagerLimit(int organisationId, int userId = 0
 
             return users >= accountManagementSettings.MaximumNumberOfAccountManagers;
         }
+
+        public bool IsPasswordValid(AspNetUser user, string email, string newPassword)
```
Comment (ghost) [inline src/NHSD.GPIT.BuyingCatalogue.Services/Users/UsersService.cs]: Glancing more at this - we shouldn't need `string email` since the `AspNetUser` is already being passed in. I propose `email` is removed and the `Where` lambda is updated to search by the `user.Id`.

Additionally, to make this method async, you could add `.ToListAsync()` after the `.Take`, and `await` the operation by placing the `await` keyword just after the `=` on line 180. The method return type would need to change to `Task<bool>` in both the interface and class. That way, the database call will be a non-blocking operation. For the method implementation in the class, `async` will need to be added before `Task<bool>`

File: src/NHSD.GPIT.BuyingCatalogue.Services/Users/UsersService.cs [hunk 10]
```diff
@@ -16,11 +21,17 @@ public sealed class UsersService : IUsersService
     {
         private readonly UserManager<AspNetUser> userManager;
         private readonly AccountManagementSettings accountManagementSettings;
+        private readonly BuyingCatalogueDbContext dbContext;
+        private readonly IPasswordHasher<AspNetUser> passwordHash;
+        private readonly PasswordResetSettings passwordRestsettings;
 
-        public UsersService(UserManager<AspNetUser> userManager, AccountManagementSettings accountManagementSettings)
+        public UsersService(UserManager<AspNetUser> userManager, AccountManagementSettings accountManagementSettings, BuyingCatalogueDbContext dbContext, IPasswordHasher<AspNetUser> passwordHash, PasswordResetSettings passwordRestsettings)
```
Comment (ghost) [inline src/NHSD.GPIT.BuyingCatalogue.Services/Users/UsersService.cs]: Spelling mistake - `passwordResetSettings`

**Review by bjss-colin — CHANGES_REQUESTED** [review]

**Review by bjss-colin — DISMISSED** [review]
Looks good, nice work 👍 

It just needs test approval and a demo to a BA prior to being merged

**Review by imranbashir01 — DISMISSED** [review]

**Review by imranbashir01 — DISMISSED** [review]

File: src/NHSD.GPIT.BuyingCatalogue.Services/Users/UsersService.cs [hunk 11]
```diff
@@ -163,5 +174,25 @@ public async Task<bool> IsAccountManagerLimit(int organisationId, int userId = 0
 
             return users >= accountManagementSettings.MaximumNumberOfAccountManagers;
         }
+
+        public async Task<bool> IsPasswordValid(AspNetUser user, string newPassword)
+        {
+            var passwordHistQuery1 = await dbContext.AspNetUsers.TemporalAll()
+                .Where(x => x.Id == user.Id)
+                .Select(x => new { x.PasswordHash, x.LastUpdated })
+                .Distinct()
```
Comment (ghost) [inline src/NHSD.GPIT.BuyingCatalogue.Services/Users/UsersService.cs]: This may also need to be updated to
```csharp
.DistinctBy(x => x.PasswordHash)
```
as the `Distinct` isn't guaranteed to return unique passwords since it now includes `LastUpdated`.

**Review by bjss-colin — CHANGES_REQUESTED** [review]

**Review by bjss-colin — APPROVED** [review]

**Review by imranbashir01 — APPROVED** [review]
